Connect Infill Polygons AND Multiply Infill

BagelOrb opened the pull request:
![image](https://user-images.githubusercontent.com/8895761/40191790-bc1590f2-5a02-11e8-9778-a371412a5daa.png)


The combination of the features:
https://github.com/Ultimaker/CuraEngine/pull/773
and
https://github.com/Ultimaker/CuraEngine/pull/767

BagelOrb commented:
This PR fixes merge conflicts and the frontend settigns allow intermixing with a new commit.

Ghostkeeper commented:
Developers, see issue CURA-5400.

Ghostkeeper commented:
I fixed the merge conflict due to the new prime tower code.

BagelOrb commented:
Yay! :tada: 

It still says "This branch cannot be rebased due to conflicts" though.
You should have done a rebase on master to prevent that....

Do you need me to do something? What's handy here?

Ghostkeeper commented:
Oh, it doesn't say that for me. It shows the button to merge and all. I don't think you need to do something now.

To give you an update, we are pushing for some of these pull requests to get merged because of the refactor coming up. A branch was already branched off for the refactor but so far the changes are limited to some new classes. It would be good to have the pull requests in before too much has changed.

BagelOrb commented:
Ok thanks for the info.

Also: I wouldn't call a 2 month wait 'pushing it'.

alekseisasin commented on src/utils/PolygonConnector.cpp:
```diff
@@ -0,0 +1,271 @@
+        result.a = *connection;
+        result.b = *second_connection;
+        // ensure that b is always the right connection and a the left
+        Point a_vec = result.a.to.p() - result.a.from.p();
```
If this constraint  "ensure that b is always the right.."  then we probably need to add to the description

BagelOrb replied:
It is already in the documentation of PolygonBridge: `--> connection a is always the left one`

BagelOrb commented on src/infill.cpp:
```diff
@@ -103,13 +130,86 @@ void Infill::generate(Polygons& result_polygons, Polygons& result_lines, const S
+    Polygons outline = in_outline.offset(outline_offset);
+
+    Polygons result;
+    Polygons first_offset = result_lines.offsetPolyLine(offset).unionPolygons(result_polygons.offset(offset).difference(result_polygons.offset(-offset)));
```
Done: c409ab3923953129146b1cd1829a4f187e2ba553

BagelOrb commented on src/utils/PolygonConnector.cpp:
```diff
@@ -0,0 +1,271 @@
+        std::optional<PolygonBridge> bridge = getBridge(current, to_connect);
+        if (bridge)
+        {
+            all_bridges.push_back(*bridge); // just for keeping scores
```
Hmm I don't know.

The tests depend on this. There are tests in the test suite which test the created bridges, but if the bridges are not separately stored there is no way to deduce from a connected polygon where the bridges were.

I don't think we should refactor the whole code so that the test case can test the code without the need of a `all_bridges`, because that would require quite a refactor while it is just for testing purposes.

One option would be to construct the class with some variable `debug_mode = true` and only then add the bridge to the list.

What do you think?

alekseisasin replied:
I think we can add then "// just for keeping scores. This is needed for test cases", 
Then it is easier to understand the reason.

BagelOrb replied:
But then the test cases won't compile!
You'll get compile errors.

I always compile the test cases... I don't want to have to change code every time I compile test cases.

alekseisasin replied:
I mean, let's keep the tack all_bridges
Only change:
all_bridges.push_back(*bridge); // just for keeping scores. This is needed for test cases.

BagelOrb replied:
Oh wait the test case doesn't fail, but half of the tests become superfluous because they don't really test what they should be testing any more.

BagelOrb commented on src/utils/PolygonConnector.cpp:
```diff
@@ -0,0 +1,271 @@
+}
+
+
+Polygon PolygonConnector::connect(const PolygonConnector::PolygonBridge& bridge)
```
Huh?!

This is the main function of the class. The sole purpose of this class is to encapsulate this function. We could rename it to `compute` or something, but that would be even less descriptive.

What do you propose?

alekseisasin replied:
ConnectBridgeParts

BagelOrb replied:
Oh wait, this wasn't the main function.

I changed it into `connectPolygonsAlongBridge`: 08f947ecabdbaaa1fef1f2c1c8b2684dc1677af1

BagelOrb commented on src/utils/PolygonConnector.cpp:
```diff
@@ -0,0 +1,271 @@
+        {
+            break;
+        }
+        first_iter = false;
```
This is just a loop to add all verts from start to end. The difficulty in reading it is caused by the fact that the processing direction can be flipped which introduces all kind of off-by-1 stuff.

There's not a lot I can add to this code to make it more clear, but I've added a small comment.

BagelOrb replied:
ebfe03a444f08eadab5eb12fb9f7dee05682df09

BagelOrb commented:
(Note: I squashed the last two commits which only added a bit of documentation. That's why it appears 3 times above. Nothing to pay attention to.)

alekseisasin commented:
[UM3_7.curaproject.3mf.zip](https://github.com/Ultimaker/CuraEngine/files/2219367/UM3_7.curaproject.3mf.zip)
Rename to UM3_7.curaproject.3mf

In Layer view, you can see that the model is missing infill.
Strange behavior after changing the Gradual Infill Steps.

BagelOrb commented:
There is some weird frontend crash when I rebase this branch on master. Nothing I can solve....

BagelOrb commented:
How did you test this branch?!

alekseisasin commented:
For CuraEngine
git fetch origin
git checkout -b connect_multiply_infill origin/connect_multiply_infill

For Cura:
git checkout -b BagelOrb-connect_multiply_infill master
git pull https://github.com/BagelOrb/Cura.git connect_multiply_infill

BagelOrb commented:
I have to update Qt, but I can't get it to work. That's unrelated to this PR tho...

BagelOrb commented:
Are you sure the strange behavior is due to this feature? Perhaps it still happens when you disable multiply infill or connect polygons...

alekseisasin commented:
I checked again and it happens if "Infill Line Multiplier" is even number. It works fine for 1, 3,5, etc

BagelOrb commented:
Could you please attach a screenshot?

Can't get Cura to run atm... :/

On Tue, Jul 24, 2018 at 5:47 PM, alekseisasin <notifications@github.com>
wrote:

> I checked again and it happens if "Infill Line Multiplier" is even number.
> It works fine for 1, 3,5, etc
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Ultimaker/CuraEngine/pull/774#issuecomment-407455139>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIe9EQ_hp5GbDPe7zHrzP77tfw-t4VS_ks5uJ0GqgaJpZM4UDauO>
> .
>



-- 

Kind regards,



Tim Kuipers

Ultimaker BV



www.ultimaker.com

BagelOrb commented:
Yay! I got Cura back on track! :tada: 

Doing the OS upgrade did help :)
Now on Linux Mint 19. I'm already a year in the future, you know ;)

alekseisasin commented:
If you can open, then I will attach images :)
![1](https://user-images.githubusercontent.com/29890821/43186189-466498cc-8fee-11e8-90f9-3719c013a8e8.png)
![2](https://user-images.githubusercontent.com/29890821/43186198-4a1bac26-8fee-11e8-8294-6ac08669bf30.png)

BagelOrb commented:
Ouch. That burned! Haha

I see the problem. Will fix it as soon as I can load projects again ;)

BagelOrb commented:
No. Actually I see in the profile that you have gradual infill enabled and the Cross pattern. Isn't it true that those two settings don't work together anyway?

alekseisasin commented:
Not only for Cross pattern. The same issue for all infill types: Grid, Lines, Triangle and etc...

BagelOrb commented:
Ah yes. I'm on it then!

BagelOrb commented:
I fixed the problem, but the way I implemented it gradual infill will still cause overlapping lines.

Solving that problem would require a whole rewrite, though.
Now the multiplication is done within the infill class, which makes it easily portable to other filling areas such as the raft.
However, gradual infill calls the infill generation multiple times, so the infill multiplication is done separately for each density region without preventing overlap with the other density regions.

I propose to accept this limitation for now.
What do you think?

BagelOrb commented:
![image](https://user-images.githubusercontent.com/8895761/43193002-30f0ae34-8fff-11e8-86ad-a11d83d488f9.png)

BagelOrb commented:
Whoops, https://github.com/Ultimaker/CuraEngine/pull/774 was the master PR which oncorporated all other PRs :O

I just had to rebase that PR on this one...

alekseisasin commented:
Ok, if we accept the limitation. Do you have some other changes?

alekseisasin commented:
Ok,  I will merge together with Cura PR: https://github.com/Ultimaker/Cura/pull/3827

BagelOrb commented:
:tada:
